WordPress/Sniffs/WP/I18nSniff.php

```diff
@@ -0,0 +1,256 @@
+			if ( ! empty( $this_token['parenthesis_closer'] ) ) {
+				for ( $j = $i + 1; $j <= $this_token['parenthesis_closer']; $j += 1 ) {
+					$tokens[ $j ]['token_index'] = $j;
+					$argument_tokens[] = $tokens[ $j ];
+				}
+				$i = $this_token['parenthesis_closer'];
+			}
+		}
+		if ( ! empty( $argument_tokens ) ) {
+			$arguments_tokens[] = $argument_tokens;
+		}
+		unset( $argument_tokens );
+
+		$argument_assertions = array();
+		if ( in_array( $translation_function, array( '__', 'esc_attr__', 'esc_html__', '_e', 'esc_attr_e', 'esc_html_e', 'translate' ) ) ) {
+			$argument_assertions[] = array( '$text', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_x', '_ex', 'esc_attr_x', 'esc_html_x', 'translate_with_gettext_context' ) ) ) {
+			$argument_assertions[] = array( '$text', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$context', 'check_literal_string_context_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_n', '_n_noop' ) ) ) {
+			$argument_assertions[] = array( '$single', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$plural', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$number', 'check_number_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_nx', '_nx_noop' ) ) ) {
+			$argument_assertions[] = array( '$single', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$plural', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$number', 'check_number_tokens' );
+			$argument_assertions[] = array( '$context', 'check_literal_string_context_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		}
+
+		$argument_stack_ptr = $func_open_paren_token;
+		foreach ( $argument_assertions as $argument_assertion ) {
+			$argument_tokens = array_shift( $arguments_tokens );
+			if ( ! $argument_tokens ) {
+				$argument_tokens = array();
+			} else {
+				$argument_stack_ptr = $argument_tokens[0]['token_index'];
+			}
+			$method_name = $argument_assertion[1];
+			$arg_name = $argument_assertion[0];
+			call_user_func( array( $this, $method_name ), $phpcs_file, $arg_name, $argument_stack_ptr, $argument_tokens );
+		}
+	}
+
+	/**
+	 * Check if supplied tokens represent a translation text string literal.
+	 *
+	 * @param PHP_CodeSniffer_File $phpcs_file The file being scanned.
+	 * @param string $arg_name
+	 * @param int $stack_ptr
+	 * @param array $tokens
+	 * @return bool
+	 */
+	protected function check_literal_string_text_tokens( $phpcs_file, $arg_name = '$text', $stack_ptr, $tokens = array() ) {
```

Comment:
> There's an unfortunate amount of duplication between check_literal_string_text_tokens, check_literal_string_context_tokens, and check_string_domain_tokens. However, an immediate refactoring wasn't apparent which would allow custom messages as well as error codes for each argument type.

Aren't the messages and error codes identical other than the arg names though? ("translatable text", "context", "text domain") I think it could be refactored to a single method that would just accept those strings and build the error messages and codes from them?